keps/sig-network/0013-20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -0,0 +1,626 @@
+  - Copy V1 PodIP to core PodIPs[0]
+- Else if only V1 PodIPs[] is provided:
+  - Copy V1 PodIPs[] to core PodIPs[]
+- Else if both V1 PodIP and V1 PodIPs[] are provided:
+  - Verify that V1 PodIP matches V1 PodIPs[0]
+  - Copy V1 PodIPs[] to core PodIPs[]
+- Delete any duplicates in core PodIPs[]
+
+##### Core (Internal) to V1 Conversion
+  - Copy core PodIPs[0] to V1 PodIP
+  - Copy core PodIPs[] to V1 PodIPs[]
+
+#### kubelet Startup Configuration for Dual-Stack Pod CIDRs
+The existing "--pod-cidr" option for the [kubelet startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) will be modified to support multiple IP CIDRs in a comma-separated list (rather than a single IP string), i.e.:
+```
+  --pod-cidr  ipNetSlice   [IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+#### kube-proxy Startup Configuration for Dual-Stack Pod CIDRs
+The existing "cluster-cidr" option for the [kube-proxy startup configuration](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-proxy/) will be modified to support multiple cluster CIDRs in a comma-separated list (rather than a single IP string), i.e:
+```
+  --cluster-cidr  ipNetSlice   [IP CIDRs, comma separated list of CIDRs, Default: []]
+```
+Only the first address of each IP family will be used; all others will be ignored.
+
+#### 'kubectl get pods -o wide' Command Display for Dual-Stack Pod Addresses
+The output for the 'kubectl get pods -o wide' command will need to be modified to display a comma-separated list of IPs for each pod, e.g.:
+```
+       kube-master# kubectl get pods -o wide
+       NAME               READY     STATUS    RESTARTS   AGE       IP                          NODE
+       nginx-controller   1/1       Running   0          20m       fd00:db8:1::2,192.168.1.3   kube-minion-1
+       kube-master#
+```
+
+#### 'kubectl describe pod ...' Command Display for Dual-Stack Pod Addresses
+The output for the 'kubectl describe pod ...' command will need to be modified to display a comma-separated list of IPs for each pod, e.g.:
+```
+       kube-master# kubectl describe pod nginx-controller
+       .
+       .
+       .
+       IPs:     fd00:db8:1::2,192.168.1.3
+       .
+       .
+       .
+```
+
+### Container Networking Interface (CNI) Plugin Considerations
+
+This feature requires the use of the [CNI Networking Plugin API version 0.3.1](https://github.com/containernetworking/cni/blob/spec-v0.3.1/SPEC.md)
+or later. The dual-stack feature requires no changes to this API.
+
+The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of Docker that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
```

Comment:
Issue 531 was a weird docker interaction that we've fixed; we no longer have a Docker version dependency.